src/NRules/NRules.Fluent/Dsl/Rule.cs

```diff
@@ -96,31 +95,23 @@ protected IRightHandSideExpression Then()
     /// </summary>
     public abstract void Define();
 
-    internal IRuleDefinition GetDefinition()
+    internal void Define(RuleBuildContext context)
     {
-        return _definition.Value;
-    }
-
-    private IRuleDefinition BuildDefinition()
-    {
-        var ruleType = GetType();
-        var metadata = new RuleMetadata(ruleType);
-        _builder.Name(metadata.Name);
-        _builder.Description(metadata.Description);
-        _builder.Tags(metadata.Tags);
-        _builder.Property(RuleProperties.ClrType, ruleType);
-
-        if (metadata.Priority.HasValue)
+        var defaultContext = _context;
+        try
         {
-            _builder.Priority(metadata.Priority.Value);
+            _context = context;
+            Define();
         }
-        if (metadata.Repeatability.HasValue)
+        finally
         {
-            _builder.Repeatability(metadata.Repeatability.Value);
+            _context = defaultContext;
         }
+    }
 
-        Define();
-
-        return _builder.Build();
+    private void InvalidateContext()
```

Comment:
@gritcsenko doesn't the name seem confusing? Invalidate means "make invalid". Probably make more sense to call something like "ValidateContext".

Reply:
My bad. Fixed